### ISSUE TITLE
stack overflow when recovering an AS

### ISSUE THREAD

**Issue by richvdh** [matrix-org/synapse#4587]
I think we aren't doing the correct thing with the defereds here:

```python
    @defer.inlineCallbacks
    def retry(self):
        try:
            txn = yield self.store.get_oldest_unsent_txn(self.service)
            if txn:
                logger.info("Retrying transaction %s for AS ID %s",
                            txn.id, txn.service.id)
                sent = yield txn.send(self.as_api)
                if sent:
                    yield txn.complete(self.store)
                    # reset the backoff counter and retry immediately
                    self.backoff_counter = 1
                    yield self.retry()
                else:
                    self._backoff()
            else:
                self._set_service_recovered()
        except Exception as e:
            logger.exception(e)
            self._backoff()
```

reported by `@swedneck`. I'm hoping they will attach some logs.

**Comment by swedneck** [comment]
My discord bridge seems to behave like @turt2live's telegram bridge, being extremely slow after restarting and taking a few minutes to go back to normal. I can't confirm whether there are a ton of transactions, however.
Logs: https://gateway.swedneck.xyz/ipfs/QmUJPPyupdWvZoGqdCWmPwBnvLW2PHLdxnAoGEG7FajLsc/synapse.log

**Comment by richvdh** [comment]
Hopefully fixed by #5885
